[Bug] Exception Ancestor must be provided for all pushes except first

<!-- Bug report best practices: https://github.com/xamarin/Xamarin.Forms/wiki/Submitting-Issues -->

### Description
Sometimes I get this exception during development but I still haven't got it once during production, should I ignore this safely?

### Steps to Reproduce

1. Build and deploy application in Debug Mode
2. Randomly get this exception

### Expected Behavior
This exception shouldn't raise

### Actual Behavior
Exception thrown causing application crash

- Version with issue: all versions I guess
- Last known good version: none

Here's the stack trace
    **Xamarin.Forms.InvalidNavigationException:** 'Ancestor must be provided for all pushes except first'

    04-07 13:46:45.378 D/Mono    (20020): DllImport attempting to load: '/system/lib64/liblog.so'.
    04-07 13:46:45.379 D/Mono    (20020): DllImport loaded library '/system/lib64/liblog.so'.
    04-07 13:46:45.379 D/Mono    (20020): DllImport searching in: '/system/lib64/liblog.so' ('/system/lib64/liblog.so').
    04-07 13:46:45.379 D/Mono    (20020): Searching for '__android_log_print'.
    04-07 13:46:45.379 D/Mono    (20020): Probing '__android_log_print'.
    04-07 13:46:45.380 D/Mono    (20020): Found as '__android_log_print'.
    04-07 13:46:45.389 I/MonoDroid(20020): UNHANDLED EXCEPTION:
    04-07 13:46:45.394 I/MonoDroid(20020): Xamarin.Forms.InvalidNavigationException: Ancestor must be provided for all pushes except first
    04-07 13:46:45.395 I/MonoDroid(20020):   at Xamarin.Forms.Internals.NavigationModel.Push (Xamarin.Forms.Page page, Xamarin.Forms.Page ancestralNav) [0x00010] in D:\a\1\s\Xamarin.Forms.Core\NavigationModel.cs:134 
    04-07 13:46:45.395 I/MonoDroid(20020):   at Xamarin.Forms.Platform.Android.AppCompat.Platform.SetPageInternal (Xamarin.Forms.Page newRoot) [0x00046] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\Platform.cs:333 
    04-07 13:46:45.395 I/MonoDroid(20020):   at Xamarin.Forms.Platform.Android.AppCompat.Platform.SetPage (Xamarin.Forms.Page newRoot) [0x000d6] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\Platform.cs:285 
    04-07 13:46:45.395 I/MonoDroid(20020):   at Xamarin.Forms.Platform.Android.FormsAppCompatActivity.InternalSetPage (Xamarin.Forms.Page page) [0x0001a] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\FormsAppCompatActivity.cs:507 
    04-07 13:46:45.395 I/MonoDroid(20020):   at Xamarin.Forms.Platform.Android.FormsAppCompatActivity.SetMainPage () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\FormsAppCompatActivity.cs:541 
    04-07 13:46:45.395 I/MonoDroid(20020):   at Xamarin.Forms.Platform.Android.FormsAppCompatActivity.OnResume () [0x00064] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\FormsAppCompatActivity.cs:401 
    04-07 13:46:45.395 I/MonoDroid(20020):   at Android.App.Activity.n_OnResume (System.IntPtr jnienv, System.IntPtr native__this) [0x00009] in <d3b924763d4a465c85b26f6e8edc8a53>:0 
    04-07 13:46:45.395 I/MonoDroid(20020):   at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.14(intptr,intptr)
    04-07 13:46:45.408 E/AppCenterCrashes(20020): Unhandled Exception from source=AndroidEnvironment
    04-07 13:46:45.408 E/AppCenterCrashes(20020): Xamarin.Forms.InvalidNavigationException: Ancestor must be provided for all pushes except first
    04-07 13:46:45.408 E/AppCenterCrashes(20020):   at Xamarin.Forms.Internals.NavigationModel.Push (Xamarin.Forms.Page page, Xamarin.Forms.Page ancestralNav) [0x00010] in D:\a\1\s\Xamarin.Forms.Core\NavigationModel.cs:134 
    04-07 13:46:45.408 E/AppCenterCrashes(20020):   at Xamarin.Forms.Platform.Android.AppCompat.Platform.SetPageInternal (Xamarin.Forms.Page newRoot) [0x00046] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\Platform.cs:333 
    04-07 13:46:45.408 E/AppCenterCrashes(20020):   at Xamarin.Forms.Platform.Android.AppCompat.Platform.SetPage (Xamarin.Forms.Page newRoot) [0x000d6] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\Platform.cs:285 
    04-07 13:46:45.408 E/AppCenterCrashes(20020):   at Xamarin.Forms.Platform.Android.FormsAppCompatActivity.InternalSetPage (Xamarin.Forms.Page page) [0x0001a] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\FormsAppCompatActivity.cs:507 
    04-07 13:46:45.408 E/AppCenterCrashes(20020):   at Xamarin.Forms.Platform.Android.FormsAppCompatActivity.SetMainPage () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\FormsAppCompatActivity.cs:541 
    04-07 13:46:45.408 E/AppCenterCrashes(20020):   at Xamarin.Forms.Platform.Android.FormsAppCompatActivity.OnResume () [0x00064] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\FormsAppCompatActivity.cs:401 
    04-07 13:46:45.408 E/AppCenterCrashes(20020):   at Android.App.Activity.n_OnResume (System.IntPtr jnienv, System.IntPtr native__this) [0x00009] in <d3b924763d4a465c85b26f6e8edc8a53>:0 
    04-07 13:46:45.408 E/AppCenterCrashes(20020):   at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.14(intptr,intptr)
    04-07 13:46:45.421 D/Mono    (20020): Requesting loading reference 5 (of 6) of /storage/emulated/0/Android/data/com.ikonbusiness.ikonapp/files/.__override__/Microsoft.AppCenter.Crashes.dll
    04-07 13:46:45.421 D/Mono    (20020): Loading reference 5 of /storage/emulated/0/Android/data/com.ikonbusiness.ikonapp/files/.__override__/Microsoft.AppCenter.Crashes.dll asmctx DEFAULT, looking for System.Core, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
    04-07 13:46:45.422 D/Mono    (20020): Assembly Ref addref Microsoft.AppCenter.Crashes[0x7c8d51bb00] -> System.Core[0x7c380cc100]: 8
    04-07 13:46:45.465 D/Mono    (20020): DllImport searching in: '__Internal' ('(null)').
    04-07 13:46:45.465 D/Mono    (20020): Searching for 'java_interop_jnienv_new_local_ref'.
    04-07 13:46:45.465 D/Mono    (20020): Probing 'java_interop_jnienv_new_local_ref'.
    04-07 13:46:45.465 D/Mono    (20020): Found as 'java_interop_jnienv_new_local_ref'.
    04-07 13:46:45.542 W/usiness.ikonap(20020): JNI RegisterNativeMethods: attempt to register 0 native methods for android.runtime.JavaProxyThrowable
    04-07 13:46:45.544 D/Mono    (20020): DllImport searching in: '__Internal' ('(null)').
    04-07 13:46:45.545 D/Mono    (20020): Searching for 'java_interop_jnienv_throw'.
    04-07 13:46:45.545 D/Mono    (20020): Probing 'java_interop_jnienv_throw'.
    04-07 13:46:45.545 D/Mono    (20020): Found as 'java_interop_jnienv_throw'.
    **Xamarin.Forms.InvalidNavigationException:** 'Ancestor must be provided for all pushes except first'

    04-07 13:46:46.355 E/mono-rt (20020): [ERROR] FATAL UNHANDLED EXCEPTION: Xamarin.Forms.InvalidNavigationException: Ancestor must be provided for all pushes except first
    04-07 13:46:46.355 E/mono-rt (20020):   at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.14(intptr,intptr)
    04-07 13:46:46.355 E/mono-rt (20020):   at (wrapper native-to-managed) Android.Runtime.DynamicMethodNameCounter.14(intptr,intptr)

--- a/Xamarin.Forms.ControlGallery.Android/FormsAppCompatActivity.cs
+++ b/Xamarin.Forms.ControlGallery.Android/FormsAppCompatActivity.cs
@@ -92,6 +92,8 @@ namespace Xamarin.Forms.ControlGallery.Android
 				return null;
 			});
 
+			DependencyService.Register<IMultiWindowService, MultiWindowService>();
+			
 			LoadApplication(_app);
 
 #if !TEST_EXPERIMENTAL_RENDERERS
@@ -101,6 +103,11 @@ namespace Xamarin.Forms.ControlGallery.Android
 				Window.SetStatusBarColor(Color.MediumPurple.ToAndroid());
 			}
 #endif
+		}
+
+		public void ReloadApplication()
+		{
+			LoadApplication(_app);
 		}
 
 		protected override void OnResume()
@@ -113,6 +120,25 @@ namespace Xamarin.Forms.ControlGallery.Android
 		public bool IsPreAppCompat()
 		{
 			return false;
+		}
+
+		[Java.Interop.Export("BackgroundApp")]
+		public void BackgroundApp()
+		{
+			Intent intent = new Intent();
+			intent.SetAction(Intent.ActionMain);
+			intent.AddCategory(Intent.CategoryHome);
+			this.StartActivity(intent);
+		}
+
+		[Java.Interop.Export("ForegroundApp")]
+		public void ForegroundApp()
+		{
+			// this only works pre API 29
+			Intent intent = new Intent(ApplicationContext, typeof(Activity1));
+			intent.SetAction(Intent.ActionMain);
+			intent.AddCategory(Intent.CategoryLauncher);
+			this.ApplicationContext.StartActivity(intent);
 		}
 	}
 }

--- a/Xamarin.Forms.ControlGallery.Android/Issue10182Activity.cs
+++ b/Xamarin.Forms.ControlGallery.Android/Issue10182Activity.cs
@@ -1,0 +1,106 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Android.App;
+using Android.Content;
+using Android.Content.PM;
+using Android.OS;
+using Android.Renderscripts;
+using Android.Runtime;
+using Android.Views;
+using Android.Widget;
+using Xamarin.Forms.Controls.Issues;
+
+namespace Xamarin.Forms.ControlGallery.Android
+{
+	[Activity(Label = "Issue10182Activity", Icon = "@drawable/icon", Theme = "@style/MyTheme",
+		MainLauncher = false, HardwareAccelerated = true,
+		ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation | ConfigChanges.ScreenLayout | ConfigChanges.SmallestScreenSize | ConfigChanges.UiMode)]
+	public class Issue10182Activity : global::Xamarin.Forms.Platform.Android.FormsAppCompatActivity
+	{
+		Activity1 _activity1;
+		protected override void OnCreate(Bundle savedInstanceState)
+		{
+			var currentApplication = Xamarin.Forms.Application.Current;
+			TabLayoutResource = Resource.Layout.Tabbar;
+			ToolbarResource = Resource.Layout.Toolbar;
+
+			base.OnCreate(savedInstanceState);
+
+			global::Xamarin.Forms.Forms.Init(this, savedInstanceState);
+			LoadApplication(new Issue10182Application());
+
+			_activity1 = (Activity1)DependencyService.Resolve<Context>();
+
+			Device.BeginInvokeOnMainThread(async () =>
+			{
+				try
+				{
+
+					await Task.Delay(1000);
+					{
+						Intent intent = new Intent(_activity1, typeof(Activity1));
+						intent.AddFlags(ActivityFlags.ReorderToFront);
+						_activity1.StartActivity(intent);
+					}
+
+					await Task.Delay(1000);
+					{
+						Intent intent = new Intent(this, typeof(Issue10182Activity));
+						intent.AddFlags(ActivityFlags.ReorderToFront);
+						StartActivity(intent);
+					}
+
+					await Task.Delay(1000);
+					{
+						Intent intent = new Intent(_activity1, typeof(Activity1));
+						intent.AddFlags(ActivityFlags.ReorderToFront);
+						_activity1.StartActivity(intent);
+					}
+				}
+				finally
+				{
+					this.Finish();
+					_activity1.ReloadApplication();
+					currentApplication.MainPage = new Issue10182.Issue10182SuccessPage();
+				}
+			});
+		}
+
+		public class Issue10182Test : ContentPage
+		{
+			public Issue10182Test()
+			{
+				Content = new StackLayout()
+				{
+					Children =
+					{
+						new Label()
+						{
+							Text = "Hold Please. Activity should vanish soon and you'll see a success label",
+							AutomationId = "Loaded"
+						}
+					}
+				};
+			}
+		}
+
+		public class Issue10182Application : Application
+		{
+			protected override void OnStart()
+			{
+				var contentPage = new Issue10182Test();
+				base.OnStart();
+				MainPage = contentPage;
+			}
+
+			protected override void OnSleep()
+			{
+				base.OnSleep();
+				MainPage = new NavigationPage(new ContentPage());
+			}
+		}
+	}
+}

--- a/Xamarin.Forms.ControlGallery.Android/MultiWindowService.cs
+++ b/Xamarin.Forms.ControlGallery.Android/MultiWindowService.cs
@@ -1,0 +1,23 @@
+﻿#if !FORMS_APPLICATION_ACTIVITY && !PRE_APPLICATION_CLASS
+
+using Android.Content;
+using Xamarin.Forms.Controls.Issues;
+using System;
+
+namespace Xamarin.Forms.ControlGallery.Android
+{
+	public class MultiWindowService : IMultiWindowService
+	{
+		public void OpenWindow(Type type)
+		{
+			if (type == typeof(Issue10182))
+			{
+				var context = DependencyService.Resolve<Context>();
+				Intent intent = new Intent(context, typeof(Issue10182Activity));
+				context.StartActivity(intent);
+			}
+		}
+	}
+}
+
+#endif

--- a/Xamarin.Forms.ControlGallery.Android/Xamarin.Forms.ControlGallery.Android.csproj
+++ b/Xamarin.Forms.ControlGallery.Android/Xamarin.Forms.ControlGallery.Android.csproj
@@ -102,7 +102,9 @@
     <Compile Include="DisposeLabelRenderer.cs" />
     <Compile Include="DisposePageRenderer.cs" />
     <Compile Include="FormsAppCompatActivity.cs" />
+    <Compile Include="Issue10182Activity.cs" />
     <Compile Include="MainApplication.cs" />
+    <Compile Include="MultiWindowService.cs" />
     <Compile Include="PerformanceTrackerRenderer.cs" />
     <Compile Include="PlatformSpecificCoreGalleryFactory.cs" />
     <Compile Include="Resources\Resource.Designer.cs" />
@@ -407,7 +409,7 @@
       <SubType>Designer</SubType>
     </AndroidResource>
   </ItemGroup>
-  <ItemGroup Condition="Exists('Properties\MapsKey.cs')">  
+  <ItemGroup Condition="Exists('Properties\MapsKey.cs')">
     <Compile Include="Properties\MapsKey.cs" />
   </ItemGroup>
   <Import Project="$(MSBuildExtensionsPath)\Xamarin\Android\Xamarin.Android.CSharp.targets" />

--- a/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue10182.cs
+++ b/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue10182.cs
@@ -18,7 +18,7 @@ namespace Xamarin.Forms.Controls.Issues
 	[Preserve(AllMembers = true)]
 	[Issue(IssueTracker.Github, 10182, "[Bug] Exception Ancestor must be provided for all pushes except first", PlatformAffected.Android, NavigationBehavior.SetApplicationRoot)]
 #if UITEST
-	[NUnit.Framework.Category(Core.UITests.UITestCategories.Github5000)]
+	[NUnit.Framework.Category(Core.UITests.UITestCategories.Github10000)]
 	[NUnit.Framework.Category(UITestCategories.LifeCycle)]
 #endif
 	public class Issue10182 : TestContentPage

--- a/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue10182.cs
+++ b/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue10182.cs
@@ -68,11 +68,11 @@ namespace Xamarin.Forms.Controls.Issues
 			}
 		}
 
-#if UITEST
+#if UITEST && __ANDROID__
 		[Test]
 		public void AppDoesntCrashWhenResettingPage()
 		{
-			RunningApp.WaitForElement("Loaded");
+			RunningApp.WaitForElement("Success");
 		}
 #endif
 	}

--- a/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue10182.cs
+++ b/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue10182.cs
@@ -1,0 +1,79 @@
+﻿using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Text;
+using Xamarin.Forms.CustomAttributes;
+using Xamarin.Forms.Internals;
+using System.Threading.Tasks;
+using System.Threading;
+
+
+#if UITEST
+using Xamarin.UITest;
+using NUnit.Framework;
+using Xamarin.Forms.Core.UITests;
+#endif
+
+namespace Xamarin.Forms.Controls.Issues
+{
+	[Preserve(AllMembers = true)]
+	[Issue(IssueTracker.Github, 10182, "[Bug] Exception Ancestor must be provided for all pushes except first", PlatformAffected.Android, NavigationBehavior.SetApplicationRoot)]
+#if UITEST
+	[NUnit.Framework.Category(Core.UITests.UITestCategories.Github5000)]
+	[NUnit.Framework.Category(UITestCategories.LifeCycle)]
+#endif
+	public class Issue10182 : TestContentPage
+	{
+		public Issue10182()
+		{
+			
+		}
+
+		protected override void Init()
+		{
+			Content = new StackLayout()
+			{
+				Children =
+				{
+					new Label()
+					{
+						Text = "Starting Activity to Test Changing Page on Resume. If success label shows up test has passed."
+					}
+				}
+			};
+
+#if !UITEST
+			Device.BeginInvokeOnMainThread(() =>
+			{
+				DependencyService.Get<IMultiWindowService>().OpenWindow(this.GetType());
+			});
+#endif
+
+		}
+
+		public class Issue10182SuccessPage : ContentPage
+		{ 
+			public Issue10182SuccessPage()
+			{
+				Content = new StackLayout()
+				{
+					Children =
+					{
+						new Label()
+						{
+							Text = "Success.",
+							AutomationId = "Success"
+						}
+					}
+				};
+			}
+		}
+
+#if UITEST
+		[Test]
+		public void AppDoesntCrashWhenResettingPage()
+		{
+			RunningApp.WaitForElement("Loaded");
+		}
+#endif
+	}
+}

--- a/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Xamarin.Forms.Controls.Issues.Shared.projitems
+++ b/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Xamarin.Forms.Controls.Issues.Shared.projitems
@@ -1399,6 +1399,7 @@
       <DependentUpon>Issue9555.xaml</DependentUpon>
     </Compile>
     <Compile Include="$(MSBuildThisFileDirectory)Issue11031.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Issue10182.cs" />
   </ItemGroup>
   <ItemGroup>
     <EmbeddedResource Include="$(MSBuildThisFileDirectory)Bugzilla22229.xaml">

--- a/Xamarin.Forms.Controls/IMultiWindowService.cs
+++ b/Xamarin.Forms.Controls/IMultiWindowService.cs
@@ -1,0 +1,11 @@
+﻿using System;
+
+
+
+namespace Xamarin.Forms.Controls.Issues
+{
+	public interface IMultiWindowService
+	{
+		void OpenWindow(Type type);
+	}
+}

--- a/Xamarin.Forms.Platform.Android/AppCompat/FormsAppCompatActivity.cs
+++ b/Xamarin.Forms.Platform.Android/AppCompat/FormsAppCompatActivity.cs
@@ -400,7 +400,7 @@ namespace Xamarin.Forms.Platform.Android
 			if (_needMainPageAssign)
 			{
 				_needMainPageAssign = false;
-
+				SettingMainPage();
 				SetMainPage();
 			}
 

--- a/Xamarin.Forms.Platform.Android/AppCompat/Platform.cs
+++ b/Xamarin.Forms.Platform.Android/AppCompat/Platform.cs
@@ -254,8 +254,11 @@ namespace Xamarin.Forms.Platform.Android.AppCompat
 
 		internal void SettingNewPage()
 		{
-			_previousNavModel = _navModel;
-			_navModel = new NavigationModel();
+			if (Page != null)
+			{
+				_previousNavModel = _navModel;
+				_navModel = new NavigationModel();
+			}
 		}
 
 		internal void SetPage(Page newRoot)

